onnxruntime/core/optimizer/fast_gelu_fusion.cc

```diff
@@ -0,0 +1,168 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/optimizer/initializer.h"
+#include "core/optimizer/fast_gelu_fusion.h"
+#include "core/optimizer/utils.h"
+#include "core/graph/graph_utils.h"
+#include "float.h"
+#include <deque>
+
+using namespace ONNX_NAMESPACE;
+using namespace onnxruntime::common;
+namespace onnxruntime {
+
+// FastGelu supports limited data types.
+static std::vector<std::string> supported_data_types{"tensor(float16)", "tensor(float)"};
+
+static bool CheckNode(const Node& node, const std::string op_name, const int32_t opset_version, const ProviderType& provider,
+  bool require_single_output=false){
+  return graph_utils::IsSupportedOptypeVersionAndDomain(node, op_name, {opset_version}) &&
+        node.GetExecutionProviderType() == provider &&
+        optimizer_utils::IsSupportedDataType(node, supported_data_types) &&
+        (!require_single_output || node.GetOutputEdgesCount() == 1);
+}
+
+Status FastGeluFusion::ApplyImpl(Graph& graph, bool& modified, int graph_level, const logging::Logger& logger) const {
+  GraphViewer graph_viewer(graph);
+  const auto& node_topology_list = graph_viewer.GetNodesInTopologicalOrder();
+
+  for (auto node_index : node_topology_list) {
+    auto* p_node = graph.GetNode(node_index);
+    if (p_node == nullptr)
+      continue;
+
+    Node& mul1_node = *p_node;
+    ORT_RETURN_IF_ERROR(Recurse(mul1_node, modified, graph_level, logger));
+
+    if (!graph_utils::IsSupportedOptypeVersionAndDomain(mul1_node, "Mul", {7}) ||
+        !graph_utils::IsSupportedProvider(mul1_node, GetCompatibleExecutionProviders()) ||
+        mul1_node.GetOutputEdgesCount() != 1 ||
+        !optimizer_utils::IsSupportedDataType(mul1_node, supported_data_types)) {
+      continue;
+    }
+
+    int32_t input_index = -1;
+    const float mul_val = 0.044715f;
+    for (auto i = 0; i < 2; i++) {
+      if (optimizer_utils::IsInitializerWithExpectedValue(graph, *(mul1_node.InputDefs()[i]), mul_val, true)){
+        input_index = i;
+        break;
+      }
+    }
+
+    if (input_index == -1) continue;
+    auto gelu_input_arg = mul1_node.InputDefs()[(input_index + 1) % 2];
+
+
+    Node& mul2_node = *graph.GetNode(mul1_node.OutputNodesBegin()->Index());
+    input_index = optimizer_utils::IndexOfNodeInput(mul2_node, *mul1_node.MutableOutputDefs()[0]);
+    if (!CheckNode(mul2_node, "Mul", {7},  mul1_node.GetExecutionProviderType(), true) ||
+        mul2_node.MutableInputDefs()[(input_index + 1) % 2]->Name() != gelu_input_arg->Name()) {
+      continue;
+    }
+
+
+    Node& add1_node = *graph.GetNode(mul2_node.OutputNodesBegin()->Index());
```

Comment:
The Gelu in model zoo [GPT2 model](https://github.com/onnx/models/blob/master/text/machine_comprehension/gpt-2/model/model.onnx) does not have different subgraph for FastGelu.

![image](https://user-images.githubusercontent.com/30328909/74393403-d7ed4e80-4dbe-11ea-94d6-6b28fdab52c7.png)

Reply:
yes it's possible. The formula can be different but when extended, they are all equivalents.  I would prefer to add this new pattern later to unblock some subsequent work. What do you think?

Reply:
I just realized I am fusing this formula: 0.5 * x * (1.0 + torch.tanh(0.7978845608028654 * x * (1.0 + 0.044715 * x * x)))
Megatron is using this to calucate. 
It is equilant to OpenAPI Gpt2 paper formula: x * 0.5 * (1.0 + tf.tanh((np.sqrt(2 / np.pi) * (x + 0.044715 * tf.pow(x, 3))))). 

I will add a todo for fuse the second formula.

Reply:
anyway I will make a change to support both today. Thanks for the comments.

Reply:
fixed.